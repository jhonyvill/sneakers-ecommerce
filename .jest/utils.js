import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import theme from "../src/Themes";

export function renderWithTheme(children) {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
}

export const productDefault = {
  id: "id-1",
  brand: "brand mock",
  title: "title mock",
  price: 250,
  discount: 0.5,
  description: "description mock",
  pictures: [
    {
      src: "image-1.jpg",
      thumb: "thumbnail-1.jpg",
      alt: "product 1",
    },
    {
      src: "image-2.jpg",
      thumb: "thumbnail-2.jpg",
      alt: "product 2",
    },
    {
      src: "image-3.jpg",
      thumb: "thumbnail-3.jpg",
      alt: "product 3",
    },
    {
      src: "image-4.jpg",
      thumb: "thumbnail-4.jpg",
      alt: "product 4",
    },
    {
      src: "image-5.jpg",
      thumb: "thumbnail-5.jpg",
      alt: "product 5",
    },
  ],
};

export const itemDefault = {
  product: { ...productDefault },
  quantity: 1,
};
