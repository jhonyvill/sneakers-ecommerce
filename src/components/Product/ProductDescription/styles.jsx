import styled from 'styled-components'

export const Container = styled.section`
  display: flex;
  flex-direction: column;
  gap: 1rem 0.8rem;
  box-sizing: border-box;
  max-width: 28.125rem;

  @media (max-width: ${({ theme }) => theme.screen.tablet}) {
    grid-template-columns: 1fr;
    padding: 0;
    justify-self: center;
    max-width: 100%;
  }
`;

export const Brand = styled.p`
  color: ${({ theme }) => theme.colors.primaryOrange};
  text-transform: uppercase;
  font-weight: 700;
  font-size: ${({ theme }) => theme.sizes.xs};
  letter-spacing: 0.15rem;

  @media (max-width: ${({ theme }) => theme.screen.largeMobile}) {
    letter-spacing: 0.1rem;
  }
`;

export const Title = styled.h1`
  color: ${({ theme }) => theme.colors.blackOpacity};
  font-weight: 700;
  font-size: ${({ theme }) => theme.sizes.xxl};
  line-height: 1.1;

  @media (max-width: ${({ theme }) => theme.screen.largeMobile}) {
    font-size: ${({ theme }) => theme.sizes.xl};
  }
`;

export const Description = styled.p`
  line-height: 1.6;
  margin-top: 1.25rem;

  @media (max-width: ${({ theme }) => theme.screen.largeMobile}) {
    font-size: ${({ theme }) => theme.sizes.s};
    line-height: 1.4;
    margin-top: 0;
  }
`;

export const PriceContainer = styled.div`
  margin: 0.625rem 0;
  align-self: flex-start;

  @media (max-width: ${({ theme }) => theme.screen.largeMobile}) {
    display: flex;
    align-items: center;
    align-self: initial;
    justify-content: space-between;
  }
`;

export const FinalPriceContainer = styled.span`
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 0 1rem;
  font-weight: 700;
`;

export const FinalPrice = styled.span`
  color: ${({ theme }) => theme.colors.blackOpacity};
  font-size: ${({ theme }) => theme.sizes.xl};

  grid-column: 1;
`;

export const Discount = styled.span`
  color: ${({ theme }) => theme.colors.primaryOrange};
  background-color: ${({ theme }) => theme.colors.paleOrange};
  padding: 0.2rem 0.5rem;
  border-radius: 6px;
  grid-column: 2;
`;

export const Price = styled.p`
  font-weight: 700;
  text-decoration: line-through;
  opacity: 0.5;
  grid-column: 1/-1;
`;
