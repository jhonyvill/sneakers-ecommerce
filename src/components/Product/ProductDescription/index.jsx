import * as S from "./styles";
import PropTypes from "prop-types";
import { useProductPrice } from "../../../hooks/useProductPrice";

const ProductDescription = ({ product }) => {
  const { price, finalPrice, discountPercent, hasDiscount } =
    useProductPrice(product);

  return (
    <S.Container aria-label="Descrição do produto">
      <S.Brand>{product.brand}</S.Brand>
      <S.Title>{product.title}</S.Title>
      <S.Description>{product.description}</S.Description>

      {hasDiscount ? (
        <S.PriceContainer>
          <S.FinalPriceContainer>
            <S.FinalPrice>${finalPrice}</S.FinalPrice>
            <S.Discount>{discountPercent}%</S.Discount>
          </S.FinalPriceContainer>

          <S.Price>${price}</S.Price>
        </S.PriceContainer>
      ) : (
        <S.FinalPrice>${price}</S.FinalPrice>
      )}
    </S.Container>
  );
};

ProductDescription.propTypes = {
  product: PropTypes.object.isRequired,
};

export default ProductDescription;
