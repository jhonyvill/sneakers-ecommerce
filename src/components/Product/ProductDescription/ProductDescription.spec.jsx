import { screen } from "@testing-library/react";
import { renderWithTheme, productDefault } from "../../../../.jest/utils";
import ProductDescription from ".";

function renderProductDescription(product = productDefault) {
  return renderWithTheme(<ProductDescription product={product} />);
}

const productWithoutDiscount = { ...productDefault, discount: 0 };

describe("<ProductDescription />", () => {
  it("should render ProductDescription component when product have discount", () => {
    renderProductDescription();

    expect(screen.getByText("50%")).toBeInTheDocument();
    expect(screen.getByText("$125.00")).toBeInTheDocument();
    expect(screen.getByText("$250.00")).toBeInTheDocument();
    expect(screen.getByText(productDefault.brand)).toBeInTheDocument();
    expect(screen.getByText(productDefault.description)).toBeInTheDocument();
    expect(screen.getByRole("heading", { name: productDefault.title })).toBeInTheDocument();
    expect(screen.getByRole("region", { name: "descrição do produto" })).toBeInTheDocument();
  });
  it("should render ProductDescription component without discount", () => {
    renderProductDescription(productWithoutDiscount);

    expect(screen.queryByText("50%")).not.toBeInTheDocument();
    expect(screen.queryByText("$125.00")).not.toBeInTheDocument();
    expect(screen.getByText("$250.00")).toBeInTheDocument();
  });
});

describe("<ProductDescription /> snapshots", () => {
  it("should match snapshot", () => {
    const { container } = renderProductDescription();
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should match snapshot of product description without discount", () => {
    const { container } = renderProductDescription(productWithoutDiscount);
    expect(container.firstChild).toMatchSnapshot();
  });
});
