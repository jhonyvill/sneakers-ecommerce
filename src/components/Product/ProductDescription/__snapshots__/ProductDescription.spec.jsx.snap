// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDescription /> snapshots should match snapshot 1`] = `
<section
  aria-label="descrição do produto"
  class="sc-bdnyFh dlhKug"
>
  <p
    class="sc-gtsqUy cHGXaI"
  >
    brand mock
  </p>
  <h1
    class="sc-dlniIP hnpDyv"
  >
    title mock
  </h1>
  <p
    class="sc-hKFymg kgwXvP"
  >
    description mock
  </p>
  <div
    class="sc-eCAqax hWGQVW"
  >
    <span
      class="sc-jSFipO foojKQ"
    >
      <span
        class="sc-gKAaef iKCHHJ"
      >
        $
        125.00
      </span>
      <span
        class="sc-iCoHzw ixzhAG"
      >
        50
        %
      </span>
    </span>
    <p
      class="sc-fujznN joVKDY"
    >
      $
      250.00
    </p>
  </div>
</section>
`;

exports[`<ProductDescription /> snapshots should match snapshot of product description without discount 1`] = `
<section
  aria-label="descrição do produto"
  class="sc-bdnyFh dlhKug"
>
  <p
    class="sc-gtsqUy cHGXaI"
  >
    brand mock
  </p>
  <h1
    class="sc-dlniIP hnpDyv"
  >
    title mock
  </h1>
  <p
    class="sc-hKFymg kgwXvP"
  >
    description mock
  </p>
  <span
    class="sc-gKAaef iKCHHJ"
  >
    $
    250.00
  </span>
</section>
`;
